hy/core/macros.hy

```diff
@@ -20,6 +20,20 @@
          ~@(interleave (repeat name) rest))
      ~name))
 
+
+(defmacro assoc [coll k1 v1 &rest other-kvs]
+  (if (odd? (len other-kvs))
+    (macro-error (last other-kvs) "`assoc` takes an odd number of arguments"))
+  (setv c (if other-kvs (gensym "c") coll))
+  `(setv
+    ~@(if other-kvs [c coll] [])
+    ~@(sum
+      (list-comp [`(get ~c ~k) v] [[k v] (zip
+        (+ (, k1) (cut other-kvs 0 None 2))
+        (+ (, v1) (cut other-kvs 1 None 2)))])
+      [])))
```

Comment:
`~@(sum (list-comp [`(get ~c ~k) v] [[k v] (cons [k1 v1] (partition other-kvs))])` should also work. I should be testing these at the repl.